### PR TITLE
Convert crafting time calculations to use 64 bit integer

### LINKED ISSUE #44542: Integer overflow is triggered when the time for crafting exceeds or equal to 1 year
# Describe the bug

Integer overflow is triggered when the time for crafting exceeds or equal one year, happens for some reason when one checks the recipe and one lacks the necessary proficiencies, the value is correct - in days and seasons - when one possesses the required proficiencies though

# Steps To Reproduce

1. Make a crafting recipe, make the crafting time equal or greater to 1 year
2. Launch the game, find the new recipe while lacking proficiencies required in that recipe
3. One can see that the crafting time value is subject to integer overflow
4. Now acquire the necessary proficiencies
5. Open the recipe again and notice that now the crafting time is displayed correctly in days and seasons

# Expected behavior

Crafting recipe is displayed properly just like in the case of possessing the proficiencies - in days and seasons - even when the crafting time in the JSON code of the recipe exceeds or equal to 1 year

# Screenshots

Bug

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1903
- Game Version: 0.E-6414-g994a2c3 [64-bit]
- Graphics Version: Tiles
- Game Language: System language []
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    <color_light_blue>BL9</color> [BL9],
    Aftershock [aftershock],
    Blaze Industries [blazeindustries],
    C.R.I.T Expansion Mod [crt_expansion],
    Magiclysm [magiclysm],
    DinoMod [DinoMod],
    Crazy Cataclysm [crazy_cataclysm],
    Mythical Martial Arts [MMA],
    Stats Through Kills [stats_through_kills],
    SpeedyDex [speedydex],
    Fuji's Military Profession Pack [fuji_mpp],
    No Fungal Monsters [No_Fungi],
    Aftershock - UndeadPeople Tileset [AftershockUDP],
    Blaze Industries - UndeadPeople Tileset [BlazeIndustriesUDP],
    C.R.I.T Expansion Mod - UndeadPeople Tileset [crt_expansionUDP],
    Crazy Cataclysm - UndeadPeople Tileset [CrazyCataclysmUDP],
    DinoMod - UndeadPeople Tileset [DinoModUDP],
    Whaleys Hair Mod (UDP Tweaks) [WHALEYHAIR_UDP],
    Magiclysm - UndeadPeople Tileset [MagiclysmUDP],
    Mythical Martial Arts - UndeadPeople Tileset [MMAUDP],
    Bionic Space Expander [Bionic_Space_Expander],
    CBM Equipment [CBM_Arms],
    Mining Mod [Mining_Mod],
    More City Locations [cityside]
]

# Additional context

The JSON file with the recipe itself [message.zip](https://github.com/CleverRaven/Cataclysm-DDA/files/5313398/message.zip)

### PR DESCRIPTION
**Pull request by RoyBerube**
#### Summary

SUMMARY: Bugfixes "Convert crafting time calculations from 32 to 64 bit integer"

#### Purpose of change

Fixes #44542

recipe.time is stored as 100 moves per second. That allows for 35.5 weeks max to be displayed before overflow.
#### Describe the solution

Change recipe.time to int64_t then change required crafting time calculations from int to int64_t.

#### Describe alternatives you've considered
Clamping some crafting calculations to 2^31 max.
#### Testing
Full test program: no errors.

Tested with looking at crafting a batch of oversized winter survivor boots. Base crafting time 4 weeks 5 days.

Before: Time estimate is correct until 7 units then estimate jumps between meaningless positive and negative numbers.

After: Time estimate is correct all the way to 50 units.

### CODE REVIEW

File: src/activity_actor.cpp [hunk 1]
```diff
@@ -1721,9 +1721,10 @@ void craft_activity_actor::do_turn( player_activity &act, Character &crafter )
 
     // Base moves for batch size with no speed modifier or assistants
     // Must ensure >= 1 so we don't divide by 0;
-    const double base_total_moves = std::max( 1, rec.batch_time( crafter, craft.charges, 1.0f, 0 ) );
+    const double base_total_moves = std::max( static_cast<int64_t>( 1 ), rec.batch_time( crafter,
+                                    craft.charges, 1.0f, 0 ) );
     // Current expected total moves, includes crafting speed modifiers and assistants
-    const double cur_total_moves = std::max( 1, rec.batch_time( crafter, craft.charges, crafting_speed,
+    const double cur_total_moves = std::max( static_cast< int64_t >( 1 ), rec.batch_time( crafter, craft.charges, crafting_speed,
```
Comment (BrettDong) [inline src/activity_actor.cpp:1727]: `astyle`:
```suggestion
    const double cur_total_moves = std::max( static_cast< int64_t >( 1 ), rec.batch_time( crafter,
                                   craft.charges, crafting_speed,
```

File: src/recipe.cpp [hunk 2]
```diff
@@ -840,7 +841,7 @@ std::string recipe::required_all_skills_string() const
 std::string recipe::batch_savings_string() const
 {
     return ( batch_rsize != 0 ) ?
-           string_format( _( "%s%% at >%s units" ), static_cast<int>( batch_rscale * 100 ), batch_rsize )
+           string_format( _( "%s%% at >%s units" ), static_cast<int64_t>( batch_rscale * 100 ), batch_rsize )
```
Comment (BrettDong) [inline src/recipe.cpp:844]: Is 64-bit integer really necessary here? `batch_rscale` is always a percentage. And string format placeholder types don't look right.
```suggestion
           string_format( _( "%d%% at >%d units" ), static_cast<int>( batch_rscale * 100 ), batch_rsize )
```

File: src/recipe.cpp [hunk 3]
```diff
@@ -94,7 +95,7 @@ int recipe::batch_time( const Character &guy, int batch, float multiplier, size_
 
     // if recipe does not benefit from batching and we have no assistants, don't do unnecessary additional calculations
     if( batch_rscale == 0.0 && assistants == 0 ) {
-        return static_cast<int>( local_time ) * batch;
+        return static_cast<int64_t>( local_time ) * batch;
```
Comment (BrettDong) [inline src/recipe.cpp:98]: I am not sure, just came up with the thought in my mind, will there be potential precision loss here, given that now it's doing 64-bit arithmetic?

Comment (RoyBerube) [reply src/recipe.cpp:98]: Do we care that much about the precision? I'm not too concerned about it.

Comment (BrettDong) [reply src/recipe.cpp:98]: Also escalating `local_time` and `total_time` to 64-bit `double` makes more sense to me.

Comment (RoyBerube) [reply src/recipe.cpp:98]: Ok. It is done.

File: src/recipe.cpp [hunk 4]
```diff
@@ -90,14 +91,14 @@ int recipe::batch_time( const Character &guy, int batch, float multiplier, size_
         multiplier = 1.0f;
     }
 
-    const float local_time = static_cast<float>( time_to_craft_moves( guy ) ) / multiplier;
+    const double local_time = static_cast<float>( time_to_craft_moves( guy ) ) / multiplier;
```
Comment (BrettDong) [inline src/recipe.cpp:94]:
```suggestion
    const double local_time = static_cast<double>( time_to_craft_moves( guy ) ) / multiplier;
```

File: src/recipe.cpp [hunk 5]
```diff
@@ -90,14 +91,14 @@ int recipe::batch_time( const Character &guy, int batch, float multiplier, size_
...
 
     // if recipe does not benefit from batching and we have no assistants, don't do unnecessary additional calculations
     if( batch_rscale == 0.0 && assistants == 0 ) {
-        return static_cast<int>( local_time ) * batch;
+        return static_cast<int64_t>( local_time ) * batch;
     }
 
-    float total_time = 0.0f;
+    double total_time = 0.0f;
```
Comment (BrettDong) [inline src/recipe.cpp:101]:
```suggestion
    double total_time = 0.0;
```

File: src/activity_actor.cpp [hunk 6]
```diff
@@ -1721,9 +1721,11 @@ void craft_activity_actor::do_turn( player_activity &act, Character &crafter )
 
     // Base moves for batch size with no speed modifier or assistants
     // Must ensure >= 1 so we don't divide by 0;
-    const double base_total_moves = std::max( 1, rec.batch_time( crafter, craft.charges, 1.0f, 0 ) );
+    const double base_total_moves = std::max( static_cast<int64_t>( 1 ), rec.batch_time( crafter,
+                                    craft.charges, 1.0f, 0 ) );
     // Current expected total moves, includes crafting speed modifiers and assistants
-    const double cur_total_moves = std::max( 1, rec.batch_time( crafter, craft.charges, crafting_speed,
+    const double cur_total_moves = std::max( static_cast< int64_t >( 1 ), rec.batch_time( crafter,
```
Comment (BrettDong) [inline src/activity_actor.cpp:1727]:
```suggestion
    const double cur_total_moves = std::max( static_cast<int64_t>( 1 ), rec.batch_time( crafter,
```

**Review by BrettDong — APPROVED** [review]
Confirmed very long time to complete does not overflow.

**Pull request merged**